File: sol/2064/2064.cpp
struct Solution {
    int required(int k, vector<int> &quantities, int m){
        int res = 0;
        for (int i = 0; i < m; ++i){
            if (quantities[i] <= k){
                return res + m - i;
            }
            res += (quantities[i] + k - 1) / k;
        }
        return res;
    }
    int minimizedMaximum(int n, vector<int>& quantities) {
        int m = quantities.size();
        ranges::sort(quantities, std::greater{});
        int s = 1, e = quantities[0];
        while (s < e){
            int k = (s + e) / 2;
            if (this->required(k, quantities, m) > n){
                s = k + 1;
            } else {
                e = k;
            }
        }
        return s;
    }
};

File: sol/2064/2064.py
class Solution:
    def minimizedMaximum(self, n: int, quantities: List[int]) -> int:
        m = len(quantities)
        quantities.sort(reverse=True)
        def required(k):
            res = 0
            for i, x in enumerate(quantities):
                if x <= k:
                    return res + m - i
                res += (x + k - 1) // k
            return res
        s = 1
        e = quantities[0]
        while s < e:
            k = (s + e) // 2
            if required(k) > n:
                s = k + 1
            else:
                e = k
        return s
